// spectacle/lib/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SPECTACLE_MODES = exports.KEYBOARD_SHORTCUTS = exports.SYSTEM_FONT = exports.DEFAULT_SLIDE_INDEX = exports.DEFAULT_SLIDE_ELEMENT_INDEX = void 0;
var DEFAULT_SLIDE_ELEMENT_INDEX = -1;
exports.DEFAULT_SLIDE_ELEMENT_INDEX = DEFAULT_SLIDE_ELEMENT_INDEX;
var DEFAULT_SLIDE_INDEX = 0;
exports.DEFAULT_SLIDE_INDEX = DEFAULT_SLIDE_INDEX;
var SYSTEM_FONT = '-apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Helvetica Neue", Helvetica, sans-serif';
exports.SYSTEM_FONT = SYSTEM_FONT;
var KEYBOARD_SHORTCUTS = {
  PRESENTER_MODE: 'mod+shift+p',
  OVERVIEW_MODE: 'mod+shift+o',
  PRINT_MODE: 'mod+shift+r',
  EXPORT_MODE: 'mod+shift+e',
  TAB_FORWARD_OVERVIEW_MODE: 'tab',
  TAB_BACKWARD_OVERVIEW_MODE: 'shift+tab',
  SELECT_SLIDE_OVERVIEW_MODE: 'enter'
};
exports.KEYBOARD_SHORTCUTS = KEYBOARD_SHORTCUTS;
var SPECTACLE_MODES = {
  DEFAULT_MODE: 'DEFAULT_MODE',
  PRESENTER_MODE: 'PRESENTER_MODE',
  OVERVIEW_MODE: 'OVERVIEW_MODE',
  PRINT_MODE: 'PRINT_MODE',
  EXPORT_MODE: 'EXPORT_MODE'
};
exports.SPECTACLE_MODES = SPECTACLE_MODES;
