// spectacle/lib/hooks/use-broadcast-channel.js
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

import { useCallback, useEffect, useRef } from 'react';
import { ulid } from 'ulid';
import { BroadcastChannel as BroadcastChannelPolyfill } from 'broadcast-channel';

var noop = function noop() {};

var BroadcastChannel = window.BroadcastChannel || BroadcastChannelPolyfill;
export default function useBroadcastChannel(channelName) {
  var onMessage = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : noop;
  var deps = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var broadcasterId = useRef(function () {
    return ulid();
  });
  var channel = useRef();
  useEffect(function () {
    channel.current = new BroadcastChannel(channelName);
    return function () {
      var _channel$current;

      (_channel$current = channel.current) === null || _channel$current === void 0 ? void 0 : _channel$current.close();
    };
  }, [channelName]);
  var postMessage = useCallback(function (type) {
    var _channel$current2;

    var payload = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var message = {
      type: type,
      payload: payload,
      meta: {
        sender: broadcasterId.current
      }
    };
    var rawMessage = JSON.stringify(message);
    (_channel$current2 = channel.current) === null || _channel$current2 === void 0 ? void 0 : _channel$current2.postMessage(rawMessage);
  }, []); // Avoid constantly modifying the 'message' listener in the effect below

  var userMessageHandlerRef = useRef(onMessage);
  useEffect(function () {
    userMessageHandlerRef.current = onMessage; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [].concat(_toConsumableArray(deps), [postMessage]));
  useEffect(function () {
    var _channel$current3;

    if (!channel.current) return;

    var messageHandler = function messageHandler(event) {
      var rawMessage = event.data;
      var message = JSON.parse(rawMessage);
      userMessageHandlerRef.current(message);
    };

    (_channel$current3 = channel.current) === null || _channel$current3 === void 0 ? void 0 : _channel$current3.addEventListener('message', messageHandler);
    return function () {
      var _channel$current4;

      (_channel$current4 = channel.current) === null || _channel$current4 === void 0 ? void 0 : _channel$current4.removeEventListener('message', messageHandler);
    };
  }, [postMessage]);
  return [postMessage, broadcasterId.current];
}
