// spectacle/es/components/fullscreen.js
var _excluded = ["size", "color"];

var _templateObject;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { forwardRef } from 'react';
import styled from 'styled-components';
import { position } from 'styled-system';
import { useToggleFullScreen } from '../hooks/use-full-screen';
import { jsx as _jsx } from "react/jsx-runtime";
var Container = styled('div')(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  ", "\n  @media print {\n    display: none;\n  }\n"])), position);
var FullScreen = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var size = _ref.size,
      color = _ref.color,
      props = _objectWithoutProperties(_ref, _excluded);

  var toggleFullScreen = useToggleFullScreen();
  return /*#__PURE__*/_jsx(Container, _objectSpread(_objectSpread({
    ref: ref,
    className: "spectacle-fullscreen-button",
    onClick: toggleFullScreen,
    style: {
      pointerEvents: 'all'
    }
  }, props), {}, {
    children: /*#__PURE__*/_jsx("svg", {
      width: size,
      height: size,
      viewBox: "0 0 512 512",
      children: /*#__PURE__*/_jsx("path", {
        fill: color,
        d: !!document.fullscreenElement || document.webkitIsFullScreen ? 'M64 371.2h76.795V448H192V320H64v51.2zm76.795-230.4H64V192h128V64h-51.205v76.8zM320 448h51.2v-76.8H448V320H320v128zm51.2-307.2V64H320v128h128v-51.2h-76.8z' : 'M396.795 396.8H320V448h128V320h-51.205zM396.8 115.205V192H448V64H320v51.205zM115.205 115.2H192V64H64v128h51.205zM115.2 396.795V320H64v128h128v-51.205z'
      })
    })
  }));
});
FullScreen.displayName = 'Fullscreen';
FullScreen.defaultProps = {
  color: '#fff',
  size: 24
};
export default FullScreen;
